source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,193 @@
+an event will be created.
+Once the host is disconnected from the DC, Vdsm will unregister from
+the 'udev' events.
+
+
+### New vdsm API
+
+New property in 'HostStats': 'faultyPathsMap' which is a mapping of faulty paths
+indexed by the device mapper GUID.
+
+
+#### Faulty Paths Map
+
+The key of the map is the device mapper GUID.
+The value is a list of faulty device mapper paths.
+
+### Vdsm multipath event listener
+
+Introduce a new Vdsm internal service for listening to multipath events:
+
+- Code interested in receiving multipath events will subscribe a callback to receive the event
+- When an event is received, all subscribers will receive the events (in another thread).
+- If nobody is registered for events, the events are dropped
+- The multipath event monitor will be started when connecting to the pool
+- The multipath event monitor will be stopped when disconnecting from the pool
+
+
+A Udev event monitor is introduced, listening to device-mapper udev events.
+The relevant events for this feature are PATH_FAILED and PATH_REINSTATED:
+
+Variable Name: DM_ACTION
+Uevent Action(s): KOBJ_CHANGE
+Type: string
+Description:
+Value: Device-mapper specific action that caused the uevent action.
+	PATH_FAILED - A path has failed.
+	PATH_REINSTATED - A path has been reinstated.
+
+- The user can disable this feature in a specific host in Vdsm configuration file
+
+In the future, the monitor can be used for other code, for example maintaining LVM filter.
+
+### Vdsm multipath statistics handler
+
+To report multipath statistics to engine, we will add a multipath statistics handler.
+
+- The handler builds a map of all the faulty paths during startup.
+- The handler subscribes to multipath events during startup
+- The handler updates the data set of the faulty paths when its callback is called:
+    - PATH_FAILED - Add the faulty path to the map according to the device GUID
+    - PATH_REINSTATED - Remove the faulty path to the map according to the device GUID
+- Statistics are provided to engine via 'Host.getStats' verb.
+
+
+### New engine API
+
+New audit logs are added:
+- Faulty multipath paths on host "HOST_NAME" on devices: "GUID", "GUID" ... 
+- No faulty multipath paths on host "HOST_NAME"
+
+## Benefit to oVirt
+
+- The user will be able to detect storage issues faster.
+
+
+## User Experience
+
+New audit logs are added with information on the device path status.
+The logs are per host.
+Since the length of the information available in the audit log is limited,
+the full information of the failing paths will be available in the logs.
+The events are not generated on "real-time", but are based on polling of
+the statistics. Only if a change in the data received is detected, a new
+audit log will be created
+
+If a single LUN goes down on all the hosts, the user will get an alarm
+for each host in the DC.
+
+If a single path goes down on one of the hosts, the user will get an alarm
+only for this host.
```

Comment:
We should consider aggregating the info on the engine side per DC. In this case we can generate an event about multipath hosts:

- Host A, B, ... have multipath devices with faulty paths
- Multipath device A, B, ... has faulty path on all hosts

Reply:
Let's start with alert per host

Reply:
About the push/pull approach - we can do maybe a mixed pull/push approach - vdsm will emit an event if there was an error detected - engine will react to this and will send a message back to get events. When the error rate goes down, or goes down below a threashold, the engine will send a message to stop sending events. Essentially a backpressure. Could be a race here that the engine will ask to don't send events and vdsm will detect an error rate again - just to take this into account

- Requierment wise - do we want users to get a general picture or with actual details, how what is failing specifically, in what rate etc.. ?

Reply:
@rgolangh On the engine side the user does not have anything to do with 1000 events when a switch goes down, so I don't see why we should work hard to get the events when there is nothing useful to do with them.

What we need is a better way to aggregate events on the engine side so instead of sending event per host (e.g 100), we will send event per DC (1), like 100 host have problematic multipath device.

In a future version we may add a way to see multipath state on a specific host or less likely state of a multipath device in the cluster (some paths are problematic on some hosts).

Anyway I think starting simple with one event per host when host multipath status changes is good enough.

Reply:
Maybe I wasn't clear, I want us to get the min event we need to fill the requirement.
But what is the requirement? I see "... to detect storage issues faster" but what should we be presenting on error? how/when does it turn back to green? how reliable should that be to users?

Reply:
We have 3 new events per host:

1. Some multipath devices are problematic on a host - the device still function, but no all paths are available - in this case the user want to troubleshoot this issue (e.g. trouble with a switch).
2. Some multipath devices are non-functional (have no valid path). This is the result of ignoring the previous alert, or a bigger issue. The host may move to non-operational due to storage monitoring issues, but it depends
3. All devices are ok on a host - back to normal operation.

The events represent multipath state reported by udev events on the host in the last monitoring cycle.

We don't plan any action on these events at this point, only log the status change. In the future we may modify the representation of the host so user can see the alarm in the host tab or maybe in the dashboard, similar to the way you see host that may need to be upgraded.